praxis-isncsci/ui
# 11 iframe adapter

Closes #11 

## Problem

Praxis Connect wants to embed the **ISNCSCI Algorithm** inside an `iFrame`. They do not want to import the library and UI components to their project so reduce the upkeep.

## Solution

We are adding an **External Message Port Provider** to bridge the communication between a container and the ISNCSCI app.

1. I have added `src/core/boundaries/iExternalMessage.provider.ts` which has the interface declaration needed in our _clean architecture_.
2. `src/app/providers/externalMessagePort.provider/externalMessagePort.provider.ts` has the implementation. It listens to the window's `message` event expecting a port to use to communicate. This way, the containing application can decide what Platform API and port to use.

## Testing

This PR is adding several infrastructure files. In order to have a meaningful Storybook story, we introduced the `App` class and `AppStore`. For these files  just make sure the code looks OK and the application structure makes sense.

There are two GitHub Actions files in this PR:
- One builds and deploys the codebase to the dev website at https://ashy-river-029cc4703-12.westeurope.3.azurestaticapps.net/
- The second file publishes StoryBook to Chromatic: https://www.chromatic.com/builds?appId=64f8d7c6e093108e99084a70

---

- [x] 1. Can register a port to communicate with the _external message port provider_

<details>

  The **External Message Port provider** will listen to the window's `message` event. The StoryBook story will instantiate a new `MessageChannel` and pass a port. The provider will then receive the notification and register the port passed. It will then print to the console that the port has been registered.

  <summary>Test case</summary>

  1. Using **Chrome** or **Edge**, avigate to https://64f8d7c6e093108e99084a70-ovilzqiyub.chromatic.com/?path=/docs/app-externalmessageportprovider--docs
  2. Open the developer console
  3. On the console, confirm the following messages are available:
      - An external message port has been registered

</details>

---

- [x] 2. Can pass exam data to the app through the _external message port provider_

<details>

  After the port has been registered, the container will be able to pass an ISNCSCI exam to the application using the action `{action: 'SET_EXAM_DATA', exam}`.

  <summary>Test case</summary>

  1. Using **Chrome** or **Edge**, avigate to https://64f8d7c6e093108e99084a70-ovilzqiyub.chromatic.com/?path=/docs/app-externalmessageportprovider--docs
  2. Open the developer console
  3. On the story, click the button `Load random exam`
  4. On the console, confirm the following messages are available:
      - Message sent to iframe
      - An exam has been received from the external message port, followed by the form data passed to the component
      - A missing implementation exception, as the code on the app side has not been written yet.

  ![image](https://github.com/praxis-isncsci/ui/assets/1294355/5acb1ca1-3e4e-46fc-b0a4-38cbaa721b0e)

</details>
